### ISSUE TITLE
Feature request: Check for classes extending AnyVal that allocate

### ISSUE THREAD

**Issue by benjumanji** [sksamuel/scapegoat#96]
Would be nice to have something prove that a value class didn't allocate. I guess you could either check for situations outlined that are known to cause allocation, or some kind of inspection on class files to check that all you have are references to static method calls. Or something. I'd be happy to have a go implementing it if you were willing to give some guidance on how it should be done.

**Comment by RichardBradley** [comment]
I was looking into similar stuff recently, and I concluded that it's pointless to check whether value classes allocate or not because the perf difference is not measurable.

See https://groups.google.com/forum/#!topic/scala-user/XdQnbcs2SRM

On our Scala projects I've changed our coding standards to suggest not adding "extends AnyVal" to extension method implicit classes because it's distracting and doesn't have any measurable benefit.

HTH

**Comment by RichardBradley** [comment]
> some kind of inspection on class files to check that all you have are references to static method calls

Scapegoat operates at the Scala language level; it doesn't currently have access to the compiled bytecode. 
You'd have to pattern match on the "situations outlined that are known to cause allocation"; you couldn't check for the actual allocation bytecode calls as things currently stand.

**Comment by sksamuel** [comment]
The groups discussion you linked to says that it is benefiical, but usually
your other computations are order of magnitude slower so it doesn't matter.
Maybe worth having it, but off by default?

**Comment by RichardBradley** [comment]
In my opinion, the distraction of worrying about whether your classes are declared properly for AnyVal allocation avoidance is more expensive than the tiny perf benefit. Also, hopefully future compilers will get better at optimising this sort of thing (not to mention the effects of the JITing optimiser), so it's not clear that any time spent optimising your code for this will survive future Scala changes.

I think there is a cost to having this kind of warning turned on: developers will spend time fixing up their classes / traits and their usage to avoid AnyVal allocations.

I can't see any harm in having it available but off by default, if other users have different opinions on the cost / benefits here.

**Comment by sksamuel** [comment]
I agree off by default. But there is a use for it so i think it should be
an option?

**Comment by RichardBradley** [comment]
Sure, I would have no objections to that (not that you need my permission anyway :-)
